File: SiriusQuality/SiriusQuality-Penman/src/f90/SiriusQuality_Penman/PotentialET_PenmanComponent.f90
MODULE Potentialet_penmanmod
    USE Priestlytaylormod
    USE Penmanmod
    IMPLICIT NONE
CONTAINS

    SUBROUTINE model_potentialet_penman(netRadiation, &
        lambdaV, &
        psychrometricConstant, &
        Alpha, &
        solarRadiation, &
        hslope, &
        ih, &
        VPDair, &
        specificHeatCapacityAir, &
        rhoDensityAir, &
        conductance, &
        evapoTranspirationPenman)
        IMPLICIT NONE
        INTEGER:: i_cyml_r
        REAL, INTENT(IN) :: netRadiation
        REAL, INTENT(IN) :: lambdaV
        REAL, INTENT(IN) :: psychrometricConstant
        REAL, INTENT(IN) :: Alpha
        REAL, INTENT(IN) :: solarRadiation
        REAL, INTENT(IN) :: hslope
        INTEGER, INTENT(IN) :: ih
        REAL, INTENT(IN) :: VPDair
        REAL, INTENT(IN) :: specificHeatCapacityAir
        REAL, INTENT(IN) :: rhoDensityAir
        REAL, INTENT(IN) :: conductance
        REAL:: evapoTranspirationPriestlyTaylor
        REAL, INTENT(OUT) :: evapoTranspirationPenman
        !- Name: PotentialET_Penman -Version: 1.0, -Time step: 1
        !- Description:
    !            * Title: PotentialET Component
    !            * Authors: SQ
    !            * Reference: None
    !            * Institution: INRAE
    !            * ExtendedDescription: https://pimlday26.sciencesconf.org/program?lang=en
    !            * ShortDescription: Potential ET (based on Penman and Priestly-Taylor)
        !- inputs:
    !            * name: netRadiation
    !                          ** description : net radiation
    !                          ** inputtype : variable
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** max : 5000.0
    !                          ** min : 0.0
    !                          ** default : 1.566
    !                          ** unit : MJ m-2 d-1
    !            * name: lambdaV
    !                          ** description : latent heat of vaporization of water
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** max : 10.0
    !                          ** min : 0.0
    !                          ** default : 2.454
    !                          ** unit : MJ kg-1
    !            * name: psychrometricConstant
    !                          ** description : psychrometric constant
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** max : 1.0
    !                          ** min : 0.0
    !                          ** default : 0.66
    !                          ** unit : 
    !            * name: Alpha
    !                          ** description : Priestley-Taylor evapotranspiration proportionality constant
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** max : 100.0
    !                          ** min : 0.0
    !                          ** default : 1.5
    !                          ** unit : 
    !            * name: solarRadiation
    !                          ** description : solar Radiation
    !                          ** inputtype : variable
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** max : 1000.0
    !                          ** min : 0.0
    !                          ** default : 3.0
    !                          ** unit : MJ m-2 d-1
    !            * name: hslope
    !                          ** description : the slope of saturated vapor pressure temperature curve at a given temperature
    !                          ** inputtype : variable
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** max : 1000.0
    !                          ** min : 0.0
    !                          ** default : 0.584
    !                          ** unit : hPa degC-1
    !            * name: ih
    !                          ** description : hour of the day if the component is hourly, -999 if the component is daily
    !                          ** inputtype : variable
    !                          ** parametercategory : constant
    !                          ** datatype : INT
    !                          ** max : 24
    !                          ** min : -999
    !                          ** default : -999
    !                          ** unit : 
    !            * name: VPDair
    !                          ** description : vapour pressure density
    !                          ** inputtype : variable
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** max : 1000.0
    !                          ** min : 0.0
    !                          ** default : 2.19
    !                          ** unit : hPa
    !            * name: specificHeatCapacityAir
    !                          ** description : Specific heat capacity of dry air
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** max : 1.0
    !                          ** min : 0.0
    !                          ** default : 0.00101
    !                          ** unit : 
    !            * name: rhoDensityAir
    !                          ** description : Density of air
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** max : 
    !                          ** min : 
    !                          ** default : 1.225
    !                          ** unit : 
    !            * name: conductance
    !                          ** description : conductance
    !                          ** inputtype : variable
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** max : 10000.0
    !                          ** min : 0.0
    !                          ** default : 598.685
    !                          ** unit : m d-1
        !- outputs:
    !            * name: evapoTranspirationPenman
    !                          ** description : evapoTranspiration of Penman Monteith
    !                          ** datatype : DOUBLE
    !                          ** variablecategory : rate
    !                          ** max : 5000.0
    !                          ** min : 0.0
    !                          ** unit : g m-2 d-1
        call model_priestlytaylor(lambdaV, netRadiation,  &
                psychrometricConstant, Alpha, solarRadiation, hslope,  &
                ih,evapoTranspirationPriestlyTaylor)
        call model_penman(VPDair, specificHeatCapacityAir,  &
                psychrometricConstant, rhoDensityAir, Alpha,  &
                evapoTranspirationPriestlyTaylor, lambdaV, hslope,  &
                conductance,evapoTranspirationPenman)
    END SUBROUTINE model_potentialet_penman

END MODULE

File: SiriusQuality/SiriusQuality-Penman/src/apsim/SiriusQuality_Penman/PotentialET_PenmanComponent.cs
using Models.Core;
using Models.Utilities;
using System; 
namespace Models.Crop2ML;
     

/// <summary>
///  PotentialET_Penman component
/// </summary>
public class PotentialET_PenmanComponent 
{

    /// <summary>
    ///  constructor of PotentialET_Penman component
    /// </summary>
    public PotentialET_PenmanComponent() {}

    //Declaration of the associated strategies
    PriestlyTaylor _PriestlyTaylor = new PriestlyTaylor();
    Penman _Penman = new Penman();

    /// <summary>
    /// Gets and sets the latent heat of vaporization of water
    /// </summary>
    [Description("latent heat of vaporization of water")] 
    [Units("MJ kg-1")] 
    public double lambdaV
    {
        get
        {
             return _PriestlyTaylor.lambdaV; 
        }
        set
        {
            _PriestlyTaylor.lambdaV = value;
            _Penman.lambdaV = value;
        }
    }

    /// <summary>
    /// Gets and sets the psychrometric constant
    /// </summary>
    [Description("psychrometric constant")] 
    [Units("")] 
    public double psychrometricConstant
    {
        get
        {
             return _PriestlyTaylor.psychrometricConstant; 
        }
        set
        {
            _PriestlyTaylor.psychrometricConstant = value;
            _Penman.psychrometricConstant = value;
        }
    }

    /// <summary>
    /// Gets and sets the Priestley-Taylor evapotranspiration proportionality constant
    /// </summary>
    [Description("Priestley-Taylor evapotranspiration proportionality constant")] 
    [Units("")] 
    public double Alpha
    {
        get
        {
             return _PriestlyTaylor.Alpha; 
        }
        set
        {
            _PriestlyTaylor.Alpha = value;
            _Penman.Alpha = value;
        }
    }

    /// <summary>
    /// Gets and sets the hour of the day if the component is hourly, -999 if the component is daily
    /// </summary>
    [Description("hour of the day if the component is hourly, -999 if the component is daily")] 
    [Units("")] 
    public int ih
    {
        get
        {
             return _PriestlyTaylor.ih; 
        }
        set
        {
            _PriestlyTaylor.ih = value;
        }
    }

    /// <summary>
    /// Gets and sets the Specific heat capacity of dry air
    /// </summary>
    [Description("Specific heat capacity of dry air")] 
    [Units("")] 
    public double specificHeatCapacityAir
    {
        get
        {
             return _Penman.specificHeatCapacityAir; 
        }
        set
        {
            _Penman.specificHeatCapacityAir = value;
        }
    }

    /// <summary>
    /// Gets and sets the Density of air
    /// </summary>
    [Description("Density of air")] 
    [Units("")] 
    public double rhoDensityAir
    {
        get
        {
             return _Penman.rhoDensityAir; 
        }
        set
        {
            _Penman.rhoDensityAir = value;
        }
    }

    /// <summary>
    /// Algorithm of PotentialET_Penman component
    /// </summary>
    public void CalculateModel(PotentialET_PenmanState s,PotentialET_PenmanState s1,PotentialET_PenmanRate r,PotentialET_PenmanAuxiliary a,PotentialET_PenmanExogenous ex)
    {
        _PriestlyTaylor.CalculateModel(s,s1, r, a, ex);
        _Penman.CalculateModel(s,s1, r, a, ex);
    }

    /// <summary>
    /// Initialization of PotentialET_Penman component
    /// </summary>
    public void Init(PotentialET_PenmanState s, PotentialET_PenmanState s1, PotentialET_PenmanRate r, PotentialET_PenmanAuxiliary a, PotentialET_PenmanExogenous ex)
    {
    }

    /// <summary>
    /// constructor copy of PotentialET_Penman component
    /// </summary>
    /// <param name="toCopy"></param>
    public PotentialET_PenmanComponent(PotentialET_PenmanComponent toCopy): this() // copy constructor 
    {
        lambdaV = toCopy.lambdaV;
        psychrometricConstant = toCopy.psychrometricConstant;
        Alpha = toCopy.Alpha;
        ih = toCopy.ih;
        specificHeatCapacityAir = toCopy.specificHeatCapacityAir;
        rhoDensityAir = toCopy.rhoDensityAir;
}
}